### ISSUE TITLE
nix develop followed by nix shell leads to strange behavior in the shell

### ISSUE THREAD

**Issue by Lurkki14** [NixOS/nix#5873]
**Describe the bug**

After using `nix shell` after `nix develop` the shell prompt gets mangled and keys no longer function as expected (eg. can't browse history with arrow keys, or use completion with tab)

![nix-shell-bug](https://user-images.githubusercontent.com/44469719/148358841-4997f962-0322-413e-86ee-716148296aef.png)

**Steps To Reproduce**

1. Enter a shell with `nix develop nixpkgs#hello`
2. When in the development shell, enter another environment with `nix shell nixpkgs#libnotify`
3. The shell should now exhibit the described unexpected behavior

**Expected behavior**

For the shell to function the same as outside development environments.

**`nix-env --version` output**
nix-env (Nix) 2.4

**Additional context**

I've tested this with a couple different packages, so it doesn't seem to be a package-specific bug. The bug also doesn't occur if you reverse the order you enter the environments (`nix develop` after `nix shell`), as shown in the image.

**Comment by thufschmitt** [comment]
That’s because running `nix develop` will set `$SHELL` to a non-interactive bash (set by the `stdenv` initialisation), and `nix shell` will use the value of `$SHELL` to create a subshell.

It’s a bit unfortunate, but also tricky to avoid systematically as `$SHELL` is the most standard way to know which shell is wanted

**Comment by tschai-yim** [comment]
This also happens when you start any editor with an integrated terminal like VSCode. Specifically for VSCode you can create an ugly workaround by using the following settings:
```json
{
  "terminal.integrated.defaultProfile.linux": "bash",
  "terminal.integrated.profiles.linux": {
    "bash": {
      "path": "/run/current-system/sw/bin/bash"
    }
  }
}
```

For SEO: nix develop vscode bash broken, nix develop bash strange

**Comment by tschai-yim** [comment]
From <https://discourse.nixos.org/t/interactive-bash-with-nix-develop-flake/15486> you can just do:
```nix
buildInputs = [ pkgs.bashInteractive ];
```
Also for me `$SHELL` doesn't actually get set.

**Comment by eclairevoyant** [comment]
Workaround no longer works, though something like

```
SHELL=$(nix build nixpkgs#bash^out --no-link --print-out-paths)/bin/bash nix shell
```

should work instead of `nix shell`.

**Comment by aldrenean** [comment]
@eclairevoyant 's suggestion didn't work for me -- `nix develop` already sets $SHELL to the nix store path for bash. What does work for me is:
```
  pkgs.mkShell {
    packages = [ ... ];
    shellHook = ''
      export SHELL="/run/current-system/sw/bin/bash"
    '';
  }
```

This sets $SHELL to the same value it has outside of the flake. I'm unsure if this workaround might cause other unforeseen problems, though.

**Comment by eclairevoyant** [comment]
> `nix develop` already sets $SHELL to the nix store path for bash.

No, it's set to the store path for _non-interactive bash_. `pkgs.bash` (on unstable) or `pkgs.bashInteractive` (on stable) is interactive bash.

**Comment by aldrenean** [comment]
I'm on unstable, there is no bashInteractive package. Using the {lib.getExe pkgs.bash} expansion in the shell hook results in exactly the same $SHELL value inside of the flake as if I don't set it. Same for manually doing SHELL=(...) nix shell.

**Comment by eclairevoyant** [comment]
The only way that's possible is if you haven't updated your nixpkgs input recently.

**Comment by janvogt** [comment]
This also happens, when using a `shell.nix` file within a project repo together with [nix-direnv](https://github.com/nix-community/nix-direnv)'s `use nix` in `.envrc`. Any call to `bash` in such a project gets the non-interactive version of bash. It's very confusing.

The root cause seems to be that nix-direnv's `use nix` uses `nix print-dev-env` instead of `nix-shell` to get the environment. Unfortunately `nix print-dev-env` does not add an interactive bash to the path generated by `$stdenv/setup` like `nix-shell` seems to be doing.
